### PR TITLE
Add stash check to name()

### LINKED ISSUE #2927: Inconsistency between .summary() and .name()
For Cubes with no identifier other than a stash code the Cube.summary() method will fall back on printing the STASH attribute, if it's present. I would suggest this is done in the CFVariableMixin.name() method instead so that this logic is consistent used more widely. I'm happy to put together a pull request if this would be useful.

### PR DESCRIPTION
**Pull request by duncanwp**
Fixes  #2927

### CODE REVIEW

File: lib/iris/_cube_coord_common.py [hunk 1]
```diff
@@ -88,12 +88,13 @@ def name(self, default='unknown'):
...
-        before falling back to the value of `default` (which itself defaults to
-        'unknown').
+        First it tries :attr:`standard_name`, then 'long_name', then
+        'var_name', then the STASH attribute before falling back to
+        the value of `default` (which itself defaults to 'unknown').
 
         """
-        return self.standard_name or self.long_name or self.var_name or default
+        return self.standard_name or self.long_name or self.var_name or \
+            self.attributes.get('STASH', default)
```
Comment (djkirkham) [inline lib/iris/_cube_coord_common.py]: At least one of the tests expects `name()` to return a string, but the `'STASH'` attribute is some other type. Wrapping it in `str()` should work.

File: lib/iris/_cube_coord_common.py [hunk 2]
```diff
@@ -88,12 +88,13 @@ def name(self, default='unknown'):
...
-        before falling back to the value of `default` (which itself defaults to
-        'unknown').
+        First it tries :attr:`standard_name`, then 'long_name', then
+        'var_name', then the STASH attribute before falling back to
+        the value of `default` (which itself defaults to 'unknown').
 
         """
-        return self.standard_name or self.long_name or self.var_name or default
+        return self.standard_name or self.long_name or self.var_name or \
+            str(self.attributes.get('STASH', default))
```
Comment (pelson) [inline lib/iris/_cube_coord_common.py]: Nit: I think I'd prefer to see the ``or`` for ``default`` so that it is easy to extend further without nesting function calls. i.e. change the line to

```
str(self.attributes.get('STASH', '')) or default
```

File: lib/iris/fileformats/netcdf.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-# (C) British Crown Copyright 2010 - 2017, Met Office
+# (C) British Crown Copyright 2010 - 2018, Met Office
```
Comment (pelson) [inline lib/iris/fileformats/netcdf.py]: These are fixed by https://github.com/SciTools/iris/pull/2932. Probably best to rebase when updating this PR.

**Comment by duncanwp** [comment]
No worries - I added tests for the STASH and default names. I put them under the Cube tests rather than the Coord ones you linked to though since I wasn't sure Coords have STASH codes

**Comment by pelson** [comment]
test failure:

```
======================================================================
FAIL: test_summary_stash (iris.tests.unit.cube.test_CubeList.TestPrint)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/scitools_iris-2.0.0rc1-py2.7.egg/iris/tests/unit/cube/test_CubeList.py", line 301, in test_summary_stash
    self.assertEqual(str(self.cubes), expected)
AssertionError: '0: m01s00i004 / (unknown)              (latitude: 3; longitude: 4)' != '0: m01s00i004                          (latitude: 3; longitude: 4)'

----------------------------------------------------------------------
```

**Comment by pelson** [comment]
Looks like some whitespace issues remain:

```
======================================================================
FAIL: test_summary_stash (iris.tests.unit.cube.test_CubeList.TestPrint)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/scitools_iris-2.1.0.dev0-py2.7.egg/iris/tests/unit/cube/test_CubeList.py", line 301, in test_summary_stash
    self.assertEqual(str(self.cubes), expected)
AssertionError: '0: m01s00i004 / (unknown)              (latitude: 3; longitude: 4)' != '0: m01s00i004 / (unknown)                          (latitude: 3; longitude: 4)'
----------------------------------------------------------------------
```

**Comment by duncanwp** [comment]
@pelson there are a couple of pp_to_cf_conversion/to_netcdf tests failing because the variable name now defaults to the stash code, like:

```
-	float unknown(depth, latitude) ;
-		unknown:_FillValue = 9.96921e+36f ;
-		unknown:um_stash_source = "m??s44i101" ;
-		unknown:ukmo__process_flags = "Mean_over_an_ensemble_of_parallel_runs Time_mean_field" ;
-		unknown:grid_mapping = "latitude_longitude" ;
-		unknown:coordinates = "forecast_period forecast_reference_time time" ;
+	float m\?\?s44i101(depth, latitude) ;
+		m\?\?s44i101:_FillValue = 9.96921e+36f ;
+		m\?\?s44i101:um_stash_source = "m??s44i101" ;
+		m\?\?s44i101:ukmo__process_flags = "Mean_over_an_ensemble_of_parallel_runs Time_mean_field" ;
+		m\?\?s44i101:grid_mapping = "latitude_longitude" ;
+		m\?\?s44i101:coordinates = "forecast_period forecast_reference_time time" ;
```

I'm not particularly happy about those var_names though. Should there be some logic in the netcdf code which rejects names with special characters, or are you OK with this new behaviour?

**Comment by pp-mo** [comment]
Looking really great, but still needs a whatsnew :bouquet: :wink:

**Comment by pelson** [comment]
> I'm not particularly happy about those var_names though. Should there be some logic in the netcdf code which rejects names with special characters, or are you OK with this new behaviour?

http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#variables

**This convention does not standardize variable names.**

However http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#_naming_conventions states:

Variable, dimension and attribute names should begin with a letter and be composed of letters, digits, and underscores. Note that this is in conformance with the COARDS conventions, but is more restrictive than the netCDF interface which allows use of the hyphen character. The netCDF interface also allows leading underscores in names, but the NUG states that this is reserved for system use


**In summary:** Yes, we need to enforce variable names when writing netcdf.

**Comment by pelson** [comment]
@duncanwp - I've pushed a change to prevent invalid CF var names in #3009.

**Comment by duncanwp** [comment]
Great, thanks - I'll rebase once that's merged.

File: lib/iris/tests/test_quickplot.py [hunk 4]
```diff
@@ -149,6 +149,7 @@ def test_contourf(self):
     def test_contourf_nameless(self):
         cube = self._small()
         cube.standard_name = None
+        cube.attributes['STASH'] = ''
```
Comment (pelson) [inline lib/iris/tests/test_quickplot.py]: ``The file header at lib/iris/tests/test_quickplot.py is out of date. The last commit was in 2018, but the copyright states it was 2016.``

😢 

I'm having a robust discussion with our legal department to get rid of the year in the copyright. Watch this space... 🙏 📖 👮

Comment (duncanwp) [reply lib/iris/tests/test_quickplot.py]: Haha, that would be great - it gets me every time! At some point I'll setup PyCharm to run the style check locally...
